### PR TITLE
Listtypes

### PR DESCRIPTION
**Pull request by btracey**
Please take a look.

### CODE REVIEW

**Comment by btracey** [comment]
I think this is right, but it's late here. I'll take another glass in the morning.

**Comment by kortschak** [comment]
>I'll take another glass in the morning.

Are you learning rhyming slang?

**Comment by btracey** [comment]
Yea, I am "glass gobbledegook" -> look

**Comment by btracey** [comment]
Okay, I think this is ready for review.

**Comment by kortschak** [comment]
Ahh. I just found out why our slang for look, "squiz", is what it is; it looks like it might have once been rhyming slang. Squizzy Taylor was a notorious gangster in the 20s in Melbourne.

File: mat/list_test.go [hunk 1]
```diff
@@ -648,45 +769,86 @@ func underlyingData(a Matrix) []float64 {
...
+// correct TriKind to be read).
 var testMatrices = []Matrix{
 	&Dense{},
-	&SymDense{},
-	NewTriDense(3, true, nil),
-	NewTriDense(3, false, nil),
-	&basicVector{},
-	&VecDense{mat: blas64.Vector{Inc: 1}},
-	&VecDense{mat: blas64.Vector{Inc: 10}},
-	&DiagDense{},
```
Comment (kortschak) [inline mat/list_test.go]: Why does the zero value go away?

Comment (btracey) [reply mat/list_test.go]: It went away because it's redundant with having an incremented and a non-incremented `Dense` type.

**Comment by btracey** [comment]
Could you elaborate on what you mean by "here"?

**Comment by btracey** [comment]
If by "here" you mean "This whole file", then I would suggest that this PR stay compact at what it is (shoring up some current inconsistencies), and then I'll write a big doc comment for the file

File: mat/list_test.go (hunk 1 above)
Comment (btracey) [reply mat/list_test.go]: Okay, sorry, fixed an issue with the increments, and now the zero value is back (this is much better)

**Comment by kortschak** [comment]
By here I mean the PR. Just add a GitHub comment about what is happening, particularly with the rearrangement of types which I don't understand without going through it in detail. And then we add that comment to the commit message when the squash happens.

**Comment by btracey** [comment]
Oh, weird, I thought I had that, sorry

```
mat: Add missing types to list test

With the recent additions of Diagonal and TriBanded, there were some matrix types that were not added to the list of testing types. There were additionally some older types that had been left out. This PR completes the set of types, their transposes, and their basic counterparts to the set of matrices. Changes are made to the makeNew etc. routines in order to support this.
```

**Comment by btracey** [comment]
The re-arrangement of types was just to make it organized which was previously pretty haphazard

File: mat/triangular.go [hunk 2]
```diff
@@ -154,7 +154,7 @@ func (t *TriDense) Dims() (r, c int) {
 // Triangle returns the dimension of t and its orientation. The returned
 // orientation is only valid when n is not zero.
 func (t *TriDense) Triangle() (n int, kind TriKind) {
-	return t.mat.N, TriKind(!t.IsZero()) && t.triKind()
+	return t.mat.N, t.triKind()
```
Comment (kortschak) [inline mat/triangular.go]: Can you add a note to the commit message about why this and the same for the banded type happens.

**Pull request merged**